[FLINK-7243][flink-formats] Add parquet input format

HuangZhenQiu opened the pull request:
## What is the purpose of the change

This pull request is to create a ParquetInputFormat, so that Flink can process files in Parquet schema. Parquet record can be read as Row, Pojo and Generic Map.

## Brief change log

  - *Add schema converter for Parquet types to Flink internal types*
  - *Add ParquetRecordRead to read Parquet record as Row*
  - *Add ParquetInputFormat that can be exteneded to convert Row to Pojo or Generic Map*


## Verifying this change

This change is already covered by existing tests, such as ParquetRecordReaderTest, ParquetSchemaConverterTest and  ParquetInputFormatTest.


## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (JavaDocs)

HuangZhenQiu commented:
@suez1224  Would you please have a look this PR?

suez1224 commented:
Thanks a lot for the PR. Please fix the minor check-style issue first. I'll take a look in the next few days.

walterddr commented on flink-formats/flink-parquet/pom.xml:
```diff
@@ -73,6 +73,11 @@ under the License.
 			<scope>provided</scope>
 		</dependency>
 
+		<dependency>
```
Seems like guava is not mandatory in your implementation.

HuangZhenQiu replied:
Used in unit test. Yeah, but is not mandatory.

walterddr commented on flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetInputFormat.java:
```diff
@@ -0,0 +1,182 @@
+		ParquetFileReader fileReader = new ParquetFileReader(inputFile, options);
+		MessageType schema = fileReader.getFileMetaData().getSchema();
+		checkSchema(schema);
+		MessageType readSchema = ParquetSchemaConverter.toParquetType(readType);
```
readSchema does not depend on anything in `open(...)`, probably can move to constructor

HuangZhenQiu replied:
The checkSchema function needs to read the parquet footer first. It is too heavy to open and read through all of the files in constructor.

walterddr commented on flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetInputFormat.java:
```diff
@@ -0,0 +1,182 @@
+	private void checkSchema(MessageType schema) {
+		RowTypeInfo rootTypeInfo = (RowTypeInfo) ParquetSchemaConverter.fromParquetType(schema);
+
+		for (int i = 0; i < fieldNames.length; ++i) {
```
Don't we need to make this checkSchema recursive for nested schemas?

HuangZhenQiu replied:
The nested comparison is handled by each type of TypInfo. For example, RowTypeInfo comparison is handled recursively in  TupleTypeInfoBase.

walterddr commented on flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/utils/ParquetSchemaConverter.java:
```diff
@@ -0,0 +1,249 @@
+					typeInfo = BasicTypeInfo.STRING_TYPE_INFO;
+					break;
+				default:
+					throw new RuntimeException("Unsupported schema: " + fieldType);
```
Probably throw `UnsupportedOperationExcpetion` ?

walterddr commented on flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/utils/ParquetSchemaConverter.java:
```diff
@@ -0,0 +1,249 @@
+						}
+						Type keyType = mapKeyValType.getType(0);
+						if (!keyType.isPrimitive()
+							|| !keyType.asPrimitiveType().getPrimitiveTypeName().equals(PrimitiveType.PrimitiveTypeName.BINARY)
```
please fix checkstyle (line too long)

HuangZhenQiu replied:
Fixed. Thanks

HuangZhenQiu commented:
@walterddr @suez1224 
Fixed Rong's comments. Please continue with the review process at your most convenient time.

fhueske commented on flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetRowInputFormat.java:
```diff
@@ -0,0 +1,82 @@
+ * A subclass of {@link ParquetInputFormat} to read from Parquet files and convert to {@link Row}.
+ * It is mainly used to integrate with table API and batch SQL.
+ */
+public class ParquetRowInputFormat extends ParquetInputFormat<Row> implements ResultTypeQueryable<Row> {
```
This PR is already quite large. A `ParquetTableSource` should be added later, IMO.

HuangZhenQiu replied:
Agree. Once this PR is merged, I will create another PR https://issues.apache.org/jira/browse/FLINK-7244.

HuangZhenQiu commented on flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/utils/ParquetSchemaConverter.java:
```diff
@@ -0,0 +1,242 @@
+	private static TypeInformation<?> convertField(final Type fieldType) {
+		TypeInformation<?> typeInfo = null;
+		if (fieldType.isPrimitive()) {
+			PrimitiveType primitiveType = fieldType.asPrimitiveType();
```
This function converts a parquet primitive type to a corresponding default Flink type. The explicitly type conversion to SqlTimeTypeInfo or other types probably can handled by users when there is a need. Otherwise, we need to bring user's conversion preference in the interface.

twalthr replied:
IMHO I think we should aim to map all Parquet types to Flink types. Otherwise the integration is not end-to-end and requires a lot of customization nevertheless. New supported formats such as ORC, Avro, or JSON support all types. Which makes the connector robust.

HuangZhenQiu commented:
@docete @fhueske 
To make a clean interface for filterable parquet input format, I needs add lots of code in this PR. After considering the size of the PR, I would like to put all of the filter pushdown in the PR of ParquetTableSource.

fhueske commented:
Hi @HuangZhenQiu, agreed. 
Let's add filter push down later. I don't think we can limit the changes to `ParquetTableSource` (the filter must be pushed down into the `InputFormat`) but that's OK.

Best, Fabian

lvhuyen commented:
Hi Zhenqiu,

I tried to read a parquet file using the code in your PR, and got the following error when reading a column of type timestamp (nullable = true).

Thanks and best regards,

```
Caused by: java.lang.IllegalArgumentException: Can not set java.math.BigInteger field ... to java.lang.String
at sun.reflect.UnsafeFieldAccessorImpl.throwSetIllegalArgumentException(UnsafeFieldAccessorImpl.java:167)
at sun.reflect.UnsafeFieldAccessorImpl.throwSetIllegalArgumentException(UnsafeFieldAccessorImpl.java:171)
at sun.reflect.UnsafeObjectFieldAccessorImpl.set(UnsafeObjectFieldAccessorImpl.java:81)
at java.lang.reflect.Field.set(Field.java:764)
at org.apache.flink.formats.parquet.ParquetPojoInputFormat.convert(ParquetPojoInputFormat.java:96)
at org.apache.flink.formats.parquet.ParquetInputFormat.nextRecord(ParquetInputFormat.java:159)
at org.apache.flink.streaming.api.functions.source.ContinuousFileReaderOperator$SplitReader.run(ContinuousFileReaderOperator.java:323)
```

lvhuyen commented:
@HuangZhenQiu 
Here is the schema of that parquet file, printed in Zeppelin.
> root
>  |-- metrics_date: timestamp (nullable = true)
>  |-- counter: long (nullable = true)
>  |-- meter: double (nullable = true)
>  |-- customer_id: string (nullable = true)
I also attach that sample file here: [https://github.com/lvhuyen/flink/blob/parquet_input_format(7243)/flink-formats/flink-parquet/src/test/resources/test.parquet](url
)

I tried to debug in IntelliJ, that column is in fact stored as primitive type int96 (not 64), and as Apache's [https://github.com/apache/parquet-mr/blob/master/parquet-column/src/main/java/org/apache/parquet/column/impl/ColumnReaderImpl.java](url), int96 is treated as a String (line 274). The way they converted from ByteArray into a String at line 393 of [https://github.com/apache/parquet-mr/blob/master/parquet-column/src/main/java/org/apache/parquet/io/api/Binary.java](url) seems to be irreversible and leads to data loss (my data has metrics_date = 2018-09-01 15:02:55.0, which was read as a bytes array of [0, 118, -95, -103, 69, 49, 0, 0, -5, -126, 37, 0]. After that line 393, I got a string with length = 12 which has the same character at 3, 4, 9, and 10th position.

HuangZhenQiu commented:
@lvhuyen 
Thanks for digging out the root cause. I guess I should pass the logic type into RowPrimitiveConverter. So that different type of data stored as Binary can be handled differently. I am working on fix for it with more test case. Thanks.

lvhuyen commented:
Thank you @HuangZhenQiu.

I created a [(quick patch),](https://github.com/lvhuyen/flink/tree/parquet_input_format(7243)) for that Timestamp conversion and my Flink job has been running well with ParquetPojoInputFormat.

One more issue I found in the current build is with the parquet file having a column of type Array of primitive. As per parquet format spec, schema for this column will have 2 layers of GroupType, while in the method _ParquetSchemaConverter.convertField(final Type fieldType)_, only one is handled. So, in the result, we'll not get an array of primitive types, but one ObjectArray which has only 1 item.
This issue is with ParquetMapInputFormat only, so I have not done anything with it in my branch.

HuangZhenQiu commented:
@lvhuyen 
Thanks for the quick patch. I think the data conversion should be handled in RowConverter. I will ship a fix tonight. For the issue of the array type in ParquetMapInputFormat, I will look a look later.

HuangZhenQiu commented:
@lvhuyen 
Last fix should resolve the problem you met in PoJoInputFormat. Would you please try it out. For primitive array handling, why the problem only happens in ParquetMapInputFormat?

lvhuyen commented:
@HuangZhenQiu Sorry for the late response. 
I was wrong. The problem should also be with PojoInputFormat as well.

HuangZhenQiu commented:
@lvhuyen 
For the Array handling issue, I figured it out. it is a List back compatibility issue. When I do internal testing at my company, there is only one type of list schema needs to be handled. Thanks for digging it out.  https://github.com/apache/parquet-format/blob/master/LogicalTypes.md#lists

I created a fix. Please have a look.

fhueske commented on flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetInputFormat.java:
```diff
@@ -0,0 +1,191 @@
+
+	protected transient ParquetRecordReader<Row> parquetRecordReader;
+
+	protected transient long recordsReadSinceLastSync;
```
this field and `lastSyncedBlock` could be private as well, right?

HuangZhenQiu replied:
Agree.

fhueske commented on flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetInputFormat.java:
```diff
@@ -0,0 +1,191 @@
+
+	protected boolean isStandard;
+
+	protected final TypeInformation[] fieldTypes;
```
Can we make the `fieldTypes` and `fieldNames` fields private and add a `protected` getter method? We've made some bad experiences exposing information to subclasses as fields and not as methods.

HuangZhenQiu replied:
Done

fhueske commented on flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetInputFormat.java:
```diff
@@ -0,0 +1,191 @@
+		this.parquetRecordReader = new ParquetRecordReader<>(new RowReadSupport(), readSchema, FilterCompat.NOOP);
+		this.parquetRecordReader.initialize(fileReader, configuration);
+		if (this.recordConsumed == null) {
+			this.recordConsumed = getRuntimeContext().getMetricGroup().counter("parquet-record-consumed");
```
record-consumed -> records-consumed

fhueske commented on flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetInputFormat.java:
```diff
@@ -0,0 +1,191 @@
+		if (this.recordConsumed == null) {
+			this.recordConsumed = getRuntimeContext().getMetricGroup().counter("parquet-record-consumed");
+		}
+		LOG.debug(String.format("Open ParquetRowInputFormat with FileInputSplit [%s]", split.getPath().toString()));
```
`ParquetRowInputFormat` -> `ParquetInputFormat`

fhueske commented on flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetInputFormat.java:
```diff
@@ -0,0 +1,191 @@
+	}
+
+	@Override
+	public void open(FileInputSplit split) throws IOException {
```
I think we have to reset `recordsReadSinchLastSync` and `lastSyncedBlock` in `open()`.

HuangZhenQiu replied:
They are reset in the next Record function.
if (parquetRecordReader.getCurrentBlock() != lastSyncedBlocker) {
...
}

fhueske commented on flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetInputFormat.java:
```diff
@@ -0,0 +1,191 @@
+			recordsReadSinceLastSync = 0;
+		}
+
+		if (parquetRecordReader.hasNextRecord()) {
```
Shouldn't this be covered by the previous `reachEnd()` check?
It looks like one of these checks could be removed or at least moved from the hot path

HuangZhenQiu replied:
Yes, you are right. The check is duplicated.

fhueske commented on flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetInputFormat.java:
```diff
@@ -0,0 +1,191 @@
+			return convert(parquetRecordReader.nextRecord());
+		}
+
+		return null;
```
This case should not happen. Can you add a log statement?

HuangZhenQiu replied:
Agree. Added

fhueske commented on flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetInputFormat.java:
```diff
@@ -0,0 +1,191 @@
+			types.add(schema.getType(readFieldName));
+		}
+
+		MessageType readSchema = new MessageType(schema.getName(), types);
```
Directly return created object

HuangZhenQiu replied:
Changed accordingly.

fhueske commented on flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetInputFormat.java:
```diff
@@ -0,0 +1,191 @@
+			String readFieldName = fieldNames[i];
+			TypeInformation<?> readFieldType = fieldTypes[i];
+			if (rootTypeInfo.getFieldIndex(readFieldName) < 0) {
+				throw new IllegalArgumentException(readFieldName + " can not be found in parquet schema");
```
Throwing an exception will immediately cancel the job. Should we add a configuration option to skip files with incompatible schema?

HuangZhenQiu replied:
Agree. Added accordingly. It provide extra convenience for users.

HuangZhenQiu commented:
@fhueske 
Thanks for reviewing this PR. I can't agree more on offering a similar experience for both input formats (Parquet and ORC). I will resolve your comments in code tonight.

Best Regards
Peter Huang

HuangZhenQiu commented on flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetInputFormat.java:
```diff
@@ -0,0 +1,191 @@
+ *
+ * @param <E> The type of record to read.
+ */
+public abstract class ParquetInputFormat<E> extends FileInputFormat<E> implements
```
Done.

HuangZhenQiu commented on flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetInputFormat.java:
```diff
@@ -0,0 +1,191 @@
+		this.readType = new RowTypeInfo(fieldTypes, fieldNames);
+		this.fieldTypes = readType.getFieldTypes();
+		this.fieldNames = readType.getFieldNames();
+		this.unsplittable = true;
```
As a Parquet file is composed of row groups with a separate metadata in footer, it is definitely splittable. It acquires to read footer of each files in advance and process each file slit with the offset info of original file. It could be separate improvement that can be done in a separate PR. How do you think?

a10y replied:
Hey @HuangZhenQiu, did you end up filing a JIRA for adding rowgroup-level splitting? Couldn't find it linked from the PR ticket.

HuangZhenQiu replied:
@a10y 
This PR is closed and merged into master. I am working on the ParquetTableSource with predicate pushdown. After that, Another PR will be added for rowgroup level splitting.

HuangZhenQiu commented on flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetInputFormat.java:
```diff
@@ -0,0 +1,191 @@
+
+	protected RowTypeInfo readType;
+
+	protected boolean isStandard;
```
Yes, It was original added to distinguish the backward compatible schema. It is not needed here.

HuangZhenQiu commented on flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetInputFormat.java:
```diff
@@ -0,0 +1,191 @@
+
+	private transient Counter recordConsumed;
+
+	protected RowTypeInfo readType;
```
Agree. I removed the duplicated readType here.

HuangZhenQiu commented on flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetInputFormat.java:
```diff
@@ -0,0 +1,191 @@
+
+	protected long lastSyncedBlock = -1L;
+
+	protected ParquetInputFormat(Path path, TypeInformation[] fieldTypes, String[] fieldNames, boolean isStandard) {
```
1) added java doc for two constructors
2) Given Path is required and multiple file path is not supported, I would prefer leave it there as AvroInputFormat, otherwise, users need to call constructor then setPath. It is not that convenient.
3) Agree. I added another constructor to take MessageType as return type.
4) I think we need. The getProducerType will be needed for table source in early stage. If we rely on schema in parquet file for field type, we can only get it after open the split for now. It is too late for SQL magics right?

HuangZhenQiu commented on flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetInputFormat.java:
```diff
@@ -0,0 +1,191 @@
+
+	protected final String[] fieldNames;
+
+	protected transient ParquetRecordReader<Row> parquetRecordReader;
```
It should be private. Changed accordingly.

HuangZhenQiu commented:
@fhueske 
Resolved most of your comments. The major blocker is probably the splittable processing of Parquet files. It will probably have a big change on the PR and more test case to cover. Given this is already a big PR, how about let me work on the improvement on another PR?

fhueske commented on flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetInputFormat.java:
```diff
@@ -0,0 +1,241 @@
+		}
+
+		// Means start to create a new block
+		if (parquetRecordReader.getCurrentBlock() != lastSyncedBlock) {
```
I think it would be good if we could move all logic to keep track of the read position and reset the read position into the `ParquetRecordReader`. That means, we do not need to keep track of read records in this class, but just fetch them from the record reader when we need to take a checkpoint.

fhueske commented on flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetInputFormat.java:
```diff
@@ -0,0 +1,241 @@
+		if (lastSyncedBlock != -1) {
+			// open and read util the record we were before
+			// the checkpoint and discard the values
+			parquetRecordReader.seek(lastSyncedBlock);
```
Can we extend the `seek` method also to skip to the right record?

fhueske commented on flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetInputFormat.java:
```diff
@@ -0,0 +1,241 @@
+
+	private transient ParquetRecordReader<Row> parquetRecordReader;
+
+	private transient long recordsReadSinceLastSync;
```
remove the counters and keep track of the read position in the `ParquetRecordReader`

fhueske commented on flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/utils/ParquetRecordReader.java:
```diff
@@ -0,0 +1,196 @@
+				try {
+					currentValue = recordReader.read();
+				} catch (RecordMaterializationException e) {
+					LOG.debug("skipping a corrupt record");
```
Add more detail to the log like file name, block number, record number and increase log level to ERROR as you did below.

Depending on the configuration, we might even want to throw an exception and cancel the job.

fhueske commented on flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/utils/ParquetRecordReader.java:
```diff
@@ -0,0 +1,196 @@
+				}
+
+				if (recordReader.shouldSkipCurrentRecord()) {
+					LOG.debug("skipping record");
```
Is this happening when a record is filtered? If yes, I don't think we should log the event, but just go on.

fhueske replied:
shouldn't we `continue` in this case and try to read the next record?

fhueske commented on flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/utils/ParquetRecordReader.java:
```diff
@@ -0,0 +1,196 @@
+			} catch (RuntimeException e) {
+				LOG.error(String.format("Can not read value at %d in block %d in file %s",
+					current, currentBlock, reader.getFile()), e);
+				return false;
```
We might want to throw an exception to cancel the job depending on the configuration.

HuangZhenQiu replied:
Agree. Added.

fhueske commented on flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/utils/ParquetRecordReader.java:
```diff
@@ -0,0 +1,196 @@
+				}
+
+				if (currentValue == null) {
+					current = totalCountLoadedSoFar;
```
Doesn't this mean that we are skipping records in the current block? Why didn't the `current == totalCountLoadedSoFar` check in previous `checkRead()` call did not work?

HuangZhenQiu replied:
Yes. It is a gate keeper of the return result of parquet record reader.

fhueske commented on flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/utils/ParquetRecordReader.java:
```diff
@@ -0,0 +1,196 @@
+
+	@CheckReturnValue(when = When.NEVER)
+	public T nextRecord() {
+		return currentValue;
```
It is mandatory that `hasNextRecord()` is called before `nextRecord()`. 
Can we change `nextRecord()` method to call `hasNextRecord()` if it wasn't called before?

HuangZhenQiu commented:
@fhueske 
Thanks for the review. Resolved all of the comments except unit tests for the checkpointing logic 
1) For the question of "instead of always reading as Row and from there converting to the other types?"

In Parquet's interface, a converter is needed for each type of result. Record can be convert to row by recursively put children in particular index, but Map has to do it with Key. To reduce code duplication, I use the row as intermediate representation. So type conversion can be put in sub class of ParquetInputFormat.

2) I will add unit test for checkpoint logic tomorrow night.

HuangZhenQiu commented:
@fhueske 
Add unit test for failure recovery logic. Please review it again after the travis check turns to green.

HuangZhenQiu commented on flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/utils/ParquetRecordReader.java:
```diff
@@ -0,0 +1,196 @@
+	public void initialize(ParquetFileReader reader, Configuration configuration) {
+		this.reader = reader;
+		FileMetaData parquetFileMetadata = reader.getFooter().getFileMetaData();
+		this.fileSchema = parquetFileMetadata.getSchema();
```
There are some duplication. I simplified it,

HuangZhenQiu commented on flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/utils/ParquetRecordReader.java:
```diff
@@ -0,0 +1,196 @@
+
+	public void seek(long syncedBlock) throws IOException {
+		PageReadStore pages = null;
+		while (syncedBlock > 0) {
```
Changed. Added test case for checkpointing logic

HuangZhenQiu commented on flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/utils/ParquetRecordReader.java:
```diff
@@ -0,0 +1,196 @@
+	private MessageType fileSchema;
+	private ReadSupport<T> readSupport;
+
+	private RecordMaterializer<T> recordMaterializer;
```
Done.

fhueske commented on flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetInputFormat.java:
```diff
@@ -0,0 +1,248 @@
+			HadoopInputFile.fromPath(new org.apache.hadoop.fs.Path(split.getPath().toUri()), configuration);
+		ParquetReadOptions options = ParquetReadOptions.builder().build();
+		ParquetFileReader fileReader = new ParquetFileReader(inputFile, options);
+		MessageType schema = fileReader.getFileMetaData().getSchema();
```
rename to `fileSchema`

fhueske commented on flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetInputFormat.java:
```diff
@@ -0,0 +1,248 @@
+			this.open(split);
+		} finally {
+			if (state.f0 != -1) {
+				// open and read util the record we were before
```
rephrase comment: "seek to the read position in the split that we were at when the checkpoint was taken."

fhueske commented on flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetInputFormat.java:
```diff
@@ -0,0 +1,248 @@
+		MessageType schema = fileReader.getFileMetaData().getSchema();
+		MessageType readSchema = getReadSchema(schema);
+
+		this.skipThisSplit = false;
```
This overrides the `skipThisSplit` values that was set by `readSchema()`.

I think there are different levels at which the schema can be checked:
* Check if `fileSchema` is a super-schema of `expectedFileSchema`
* Check if `readSchema` (which might be a subschema of `expectedFileSchema`) can be read from `fileSchema`

fhueske commented on flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetInputFormat.java:
```diff
@@ -0,0 +1,248 @@
+		}
+
+		LOG.debug(String.format("Open ParquetInputFormat with FileInputSplit [%s]", split.getPath().toString()));
+		if (skipThisSplit) {
```
This check can be done after we tried to get the `readSchema`.

HuangZhenQiu commented:
@fhueske 
Thanks for your patient review. It is pretty helpful to make the PR more readable and flawless. Resolved your comments. Please read one more round at your most convenient time.

HuangZhenQiu commented:
@fhueske 
Thanks for your patient review. It is pretty helpful to make the PR more readable and flawless. Resolved your comments. Please read one more round at your most convenient time.

HuangZhenQiu commented:
@fhueske 
Thanks for your patient review. It is pretty helpful to make the PR more readable and flawless. Resolved your comments. Please read one more round at your most convenient time.

HuangZhenQiu commented:
@fhueske 
Thanks for your patient review. It is pretty helpful to make the PR more readable and flawless. Resolved your comments. Please read one more round at your most convenient time.

fhueske commented on flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetInputFormat.java:
```diff
@@ -0,0 +1,283 @@
+	/**
+	 * The config parameter which defines whether to skip file split with wrong schema.
+	 */
+	public static final String SKIP_WRONG_SCHEMA_SPLITS = "skip.splits.wrong.schema";
```
prefix with `parquet` to avoid collisions with other formats.

HuangZhenQiu commented:
@fhueske 
Resolved all of the comments except the one for timestamp rewrite. It is needed for time field of window functions. Do you prefer to use timestamp udf SQL directly in this case?

HuangZhenQiu commented:
@fhueske 
I removed timestamp override, and also update the failure recovery test case to test recovery reading file with 10 row group. Please review it as your most convenient time.

fhueske commented on flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/utils/ParquetSchemaConverter.java:
```diff
@@ -0,0 +1,359 @@
+		Type fieldType = null;
+
+		Type.Repetition repetition = inheritRepetition == null ? Type.Repetition.OPTIONAL : inheritRepetition;
+		if (typeInfo.isBasicType()) {
```
Better check with `instanceOf`

fhueske commented on flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/utils/ParquetSchemaConverter.java:
```diff
@@ -0,0 +1,359 @@
+		if (typeInfo.isBasicType()) {
+			BasicTypeInfo basicTypeInfo = (BasicTypeInfo) typeInfo;
+			if (basicTypeInfo.equals(BasicTypeInfo.BIG_DEC_TYPE_INFO)) {
+				fieldType = Types.primitive(PrimitiveType.PrimitiveTypeName.DOUBLE, repetition).named(fieldName);
```
Should be mapped to `DECIMAL` type instead of imprecise `DOUBLE`?

fhueske commented on flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/utils/ParquetSchemaConverter.java:
```diff
@@ -0,0 +1,359 @@
+			if (basicTypeInfo.equals(BasicTypeInfo.BIG_DEC_TYPE_INFO)) {
+				fieldType = Types.primitive(PrimitiveType.PrimitiveTypeName.DOUBLE, repetition).named(fieldName);
+			} else if (basicTypeInfo.equals(BasicTypeInfo.BIG_INT_TYPE_INFO)) {
+				fieldType = Types.primitive(PrimitiveType.PrimitiveTypeName.INT96, repetition).named(fieldName);
```
Map to `DECIMAL`?

fhueske commented on flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/utils/ParquetSchemaConverter.java:
```diff
@@ -0,0 +1,359 @@
+			} else if (basicTypeInfo.equals(BasicTypeInfo.SHORT_TYPE_INFO)) {
+				fieldType = Types.primitive(PrimitiveType.PrimitiveTypeName.INT32, repetition).named(fieldName);
+			} else if (basicTypeInfo.equals(BasicTypeInfo.BYTE_TYPE_INFO)) {
+				fieldType = Types.primitive(PrimitiveType.PrimitiveTypeName.BINARY, repetition).named(fieldName);
```
Map to `INT32`?

fhueske commented on flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/utils/ParquetSchemaConverter.java:
```diff
@@ -0,0 +1,359 @@
+			} else if (basicTypeInfo.equals(BasicTypeInfo.BYTE_TYPE_INFO)) {
+				fieldType = Types.primitive(PrimitiveType.PrimitiveTypeName.BINARY, repetition).named(fieldName);
+			} else if (basicTypeInfo.equals(BasicTypeInfo.CHAR_TYPE_INFO)) {
+				fieldType = Types.primitive(PrimitiveType.PrimitiveTypeName.BINARY, repetition).named(fieldName);
```
Should this be `BINARY` (`UTF8`) or `INT32`?

fhueske commented on flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/utils/ParquetSchemaConverter.java:
```diff
@@ -0,0 +1,359 @@
+		return typeInfo;
+	}
+
+	private static Type convertField(String fieldName, TypeInformation<?> typeInfo,
```
I assume only the physical type is important here?
Or do we need to be precise and provide the logical type as well because the schema is stored in the file?

HuangZhenQiu replied:
As suggested by Timo, this function considers both physical type and logical type. As you can see, it firstly convert type by physical type, if there are multiple logical types using the same physical data size to store (for example TIME_MICROS, TIMESTAMP_MICROS are used as INT64).

fhueske commented on flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/utils/ParquetSchemaConverter.java:
```diff
@@ -0,0 +1,359 @@
+		} else if (typeInfo instanceof MapTypeInfo) {
+			MapTypeInfo mapTypeInfo = (MapTypeInfo) typeInfo;
+			fieldType = Types.optionalMap()
+				.key(convertField(MAP_KEY, mapTypeInfo.getKeyTypeInfo(), Type.Repetition.REQUIRED, isStandard))
```
Check that key is a string value? Or are any kind of keys supported?

fhueske commented on flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/utils/ParquetSchemaConverter.java:
```diff
@@ -0,0 +1,359 @@
+	}
+
+	private static Type convertField(String fieldName, TypeInformation<?> typeInfo,
+									Type.Repetition inheritRepetition, boolean isStandard) {
```
rename `isStandard` to `legacyMode`?

fhueske commented on flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/utils/ParquetSchemaConverter.java:
```diff
@@ -0,0 +1,359 @@
+					.named(LIST_ARRAY_TYPE)
+					.as(OriginalType.LIST).named(fieldName);
+			}
+		} else {
```
It seems that the time types are not handled. They do not belong to the basic types.

fhueske commented on flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/utils/ParquetSchemaConverter.java:
```diff
@@ -0,0 +1,359 @@
+		return fieldType;
+	}
+
+	private boolean isElementType(Type repeatedType, String parentName) {
```
does not seem to be used. Remove?

HuangZhenQiu commented:
@fhueske 
Thanks for pointing out so many missing parts. As you pointed out, I added SqlTypeInfo conversion and enforce List and Map schema convention in Diff. Please review it when you have time.

HuangZhenQiu commented:
@fhueske I refined test cases. Would you please take a last round of review?

fhueske commented on flink-formats/flink-parquet/src/test/java/org/apache/flink/formats/parquet/utils/TestUtil.java:
```diff
@@ -0,0 +1,293 @@
+	private AvroSchemaConverter converter = new AvroSchemaConverter();
+
+	@Ignore
+	public void testSimpleSchemaConversion() {
```
Why are this and the next test ignored?
Can they be removed?

fhueske commented on flink-formats/flink-parquet/src/test/java/org/apache/flink/formats/parquet/utils/TestUtil.java:
```diff
@@ -0,0 +1,293 @@
+
+	/**
+	 * This test case is using the parquet native avro schema converter that convert
+	 * avro map value typed to required. But komodo write parquet files with map value
```
What is komodo? ;-)

fhueske commented on flink-formats/flink-parquet/src/test/java/org/apache/flink/formats/parquet/utils/TestUtil.java:
```diff
@@ -0,0 +1,293 @@
+		return t;
+	}
+
+	public static void main(String[] args) throws IOException {
```
What is the `main` method used for? Can it be removed?

HuangZhenQiu commented:
@fhueske 
Thanks for the review. I removed all of unused fields, main function and test cases. To have better code coverage, I also added test cases for projected selection for each subclass of ParquetInputFormat.

fhueske commented:
Thanks for the update @HuangZhenQiu!
I'll do another pass (and maybe even merge it) tomorrow. 

I tried to rebase the PR to the current master, but the build failed due to some unresolved dependency.
Can you rebase the PR to the current master and update the PR?

Thanks, Fabian

HuangZhenQiu commented:
@fhueske 
I finished rebase lastest upstream master just now. The compile error probably comes from the generated avro classes are committed in last several diffs. I removed them also. Thanks for the your effort of reviewing this PR.

fhueske commented:
Thanks for the update @HuangZhenQiu.
I'll do some final minor changes and will merge the PR.
Thanks for your patience and endurance for this contribution.

Best, Fabian

fpompermaier commented on flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/utils/ParquetSchemaConverter.java:
```diff
@@ -0,0 +1,437 @@
+ * Schema converter converts Parquet schema to and from Flink internal types.
+ */
+public class ParquetSchemaConverter {
+	private static final Logger LOGGER = LoggerFactory.getLogger(RowConverter.class);
```
Probably this should be LoggerFactory.getLogger(ParquetSchemaConverter.class)

lvhuyen commented on flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetPojoInputFormat.java:
```diff
@@ -0,0 +1,133 @@
+	public void open(FileInputSplit split) throws IOException {
+		super.open(split);
+		pojoFields = new Field[getFieldNames().length];
+		LOG.error("Fields number is %d", getFieldNames().length);
```
Should this be LOG.debug?

lvhuyen commented on flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetPojoInputFormat.java:
```diff
@@ -0,0 +1,133 @@
+	/**
+	 * Extracts the {@link TypeInformation}s  from {@link PojoTypeInfo} according to the given field name.
+	 */
+	private static <E> TypeInformation<?>[] extractTypeInfos(PojoTypeInfo<E> pojoTypeInfo, String[] fieldNames) {
```
This private method extractTypeInfos(...) is not used anywhere within this ParquetPojoInputFormat class. Is it redundant and should be removed?
